### PR TITLE
feat:[CDS-70330]: encryption TF plan changes for NG

### PR DESCRIPTION
**Pull request by sourabh-awashti**
## Describe your changes
These changes are related to encryption and decryption of TF plan on manager side through API calls for NG
Tests Done: ( All tests are done in NG )
Test plan
https://docs.google.com/spreadsheets/d/1AlleguoCSfOBfAC7aDsSgR77q4YkRhG0aUCfsiONuVs/edit#gid=0
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48986)

### CODE REVIEW

**Review by vlica-harness — CHANGES_REQUESTED** [review]

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java [hunk 1]
```diff
@@ -638,9 +639,24 @@ public EncryptionConfig getEncryptionConfig(Ambiance ambiance, TerraformPlanStep
...
-        identifierRef.getAccountIdentifier(), identifierRef.getOrgIdentifier(), identifierRef.getProjectIdentifier(),
-        identifierRef.getIdentifier(), false));
+    SecretManagerConfig secretManagerConfig =
+        SecretManagerConfigMapper.fromDTO(secretManagerClientService.getSecretManager(
+            identifierRef.getAccountIdentifier(), identifierRef.getOrgIdentifier(),
+            identifierRef.getProjectIdentifier(), identifierRef.getIdentifier(), false));
+    if (cdFeatureFlagHelper.isEnabled(
+            AmbianceUtils.getAccountId(ambiance), CDS_TERRAFORM_TERRAGRUNT_PLAN_ENCRYPTION_ON_MANAGER_NG)
+        && isHarnessSecretManager(secretManagerConfig)) {
+      secretManagerConfig.maskSecrets();
```
Comment (vlica-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java:649]: @sourabh-awashti 
how do we handle secretManagerConfig with maskedSecrets if there is an older delegate?

Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java:649]: We have to ask all customers who are using terraform or terragrunt to upgrade to new delegate

**Comment by sourabh-awashti** [comment]
> Can we have please added unit tests for manager and delegate?

I have added it in separate PR, will share it

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java [hunk 2]
```diff
@@ -638,9 +639,24 @@ public EncryptionConfig getEncryptionConfig(Ambiance ambiance, TerraformPlanStep
...
+    }
+    return secretManagerConfig;
+  }
+
+  public boolean isHarnessSecretManager(SecretManagerConfig secretManagerConfig) {
+    if (secretManagerConfig != null) {
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformStepHelper.java]: done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformApplyStep.java [hunk 3]
```diff
@@ -256,6 +258,9 @@ private TaskRequest obtainInheritedTask(
                     : inheritOutput.getEnvironmentVariables())
             .timeoutInMillis(
                 StepUtils.getTimeoutMillis(stepElementParameters.getTimeout(), TerraformConstants.DEFAULT_TIMEOUT))
+            .encryptDecryptPlanForHarnessSMOnManager(
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformApplyStep.java:259]: done

File: 400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResourceNG.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+import org.jetbrains.annotations.NotNull;
+import retrofit.http.Body;
+
+@Api("/encryption-ng")
+@Path("/encryption-ng")
+@Produces("application/json")
+@Slf4j
+@OwnedBy(CDP)
+public class DelegateManagerEncryptionDecryptionHarnessSMResourceNG {
+  @Inject DelegateManagerEncryptionDecryptionHarnessSMServiceNGImpl encryptionDecryptionService;
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 400-rest/src/main/java/software/wings/resources/DelegateManagerEncryptionDecryptionHarnessSMResourceNG.java:40]: done

File: 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformApplyStep.java [hunk 5]
```diff
@@ -256,6 +256,8 @@ private TaskRequest obtainInheritedTask(
...
+                helper.tfPlanEncryptionOnManager(accountId, inheritOutput.encryptionConfig))
```
[the remark this answers is not in the document]
Comment (sourabh-awashti) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/provision/terraform/TerraformApplyStep.java]: done

**Review by yogesh-chauhan — APPROVED** [review]
> Reviewed 7 of 36 files at r1, 24 of 33 files at r2, 7 of 8 files at r3, 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48986)* status: 40 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/48986)* status: 40 file(s) to review. 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained.

**Comment by sourabh-awashti** [comment]
trigger uts

**Comment by sourabh-awashti** [comment]
trigger review

**Pull request merged**